Add scatterplots to C pages

leburnett opened the pull request:
Fixes #78 

- Adds interactive scatterplots to the "C" pages of types with synapses within ME, LO or LOP. 
- The metrics are the mean of the values calculated individually for each optic lobe side (ME(R) and ME(L) etc.)
- When the page loads, the marker that represents the given type should automatically be highlighted. 
- A marker will only be shown if that type has > 50 synapses (across all cells) and >5% of its total synapse number within that region. This is why some types have eyemap plots that are not empty for certain regions but they don't have a marker in the corresponding scatterplot. 
- The cell type's marker will remain "highlighted" as you hover over other markers within the plot, but will reduce back to normal size if hovered over. 

- One downside to how these plots are currently generated is that they will only be "complete" if pages for all types (or optic lobe types) have been generated (e.g. using `pixi run neuview generate-all`).
- The plots will be made if you run `pixi run subset-medium` etc. However, the scatterplots will only have a few markers representing the cell types that have pages.

leburnett committed:
test code for new service

leburnett committed:
Merge branch 'main' into issue_78_part_2

leburnett committed:
Create scatterplot_service.py

leburnett committed:
Merge branch 'main' into issue_78_part_2

leburnett committed:
test

leburnett committed:
Merge branch 'main' into issue_78_part_2

leburnett committed:
add testing code

leburnett committed:
Update test_scatter.py

leburnett committed:
Update test_scatter.py

leburnett committed:
Merge branch 'main' into issue_78_part_2

leburnett committed:
Update cache_service.py

Update generation of cache file to include "incl_scatter" within the "roi_summary" part only for "ME", "LO" and "LOP". Uses the strict threshold service of 5% - either pre or post percentage of synapses must be >5% and there must be >50 synapses in the ROI to be included in the scatterplots.

leburnett committed:
Update scatterplot_service.py

leburnett committed:
Update generate-all workflow to incl scatters

leburnett committed:
Update scatterplot.svg.jinja

leburnett committed:
update neuron page templates for scatterplots

leburnett committed:
Update scatterplots.html.jinja

leburnett committed:
Update scatterplot.svg.jinja

Move selected marker on top of data

leburnett committed:
Create scatter_config.py

leburnett committed:
Update scatterplot_service.py

leburnett committed:
Bump version to v2.7.8

leburnett committed:
Update cli.py

leburnett committed:
some updates

leburnett committed:
Update pyproject.toml

leburnett committed:
Update scatterplot_service.py

leburnett committed:
Update scatter_config.py

leburnett committed:
update scatter templates svg and html

leburnett committed:
Remove output dir from command

leburnett committed:
Move config

leburnett committed:
Update scatterplot_service.py

leburnett committed:
update templates

leburnett committed:
Update neuron-page.js

leburnett committed:
make plots square

leburnett committed:
Remove download button for now.

leburnett committed:
Delete test_scatter.py

leburnett committed:
Delete my_file.pkl

leburnett committed:
Merge branch 'main' into issue_78_part_2

leburnett committed:
Update neuron_page_scripts.html.jinja

leburnett committed:
Update pyproject.toml

leburnett committed:
Bump version to v2.7.9

leburnett committed:
Explicitly search for neuron name

leburnett committed:
Update pyproject.toml

leburnett committed:
Update cache_service.py

leburnett committed:
Remove plot_data function from index service

leburnett committed:
remove test, update config

leburnett committed:
black formatting

leburnett committed:
Update pyproject.toml

leburnett committed:
Update pixi.lock to match pyproject.toml

leburnett committed:
Merge branch 'main' into issue_78_part_2

leburnett committed:
Update scatterplot.svg.jinja

leburnett committed:
Update pixi.lock

leburnett commented:
I will take a look into this today.

leburnett committed:
Remove exit when no data found for scatter

leburnett committed:
Merge branch 'main' into issue_78_part_2

leburnett committed:
Update pixi.lock

floesche commented:
Running `pixi run subset-medium` only shows 2 of the 3 scatter plots:

<img width="1750" height="597" alt="image" src="https://github.com/user-attachments/assets/dc9bda53-49c9-4117-9fa6-10f15c88a865" />

leburnett commented:
Yes, this is because the only OL cell type in "subset medium" is Tm3 which only has synapses within ME and LO. As mentioned above, we can extend the medium subset or makes an OL subset with more types so that you can see more points within the scatterplots and a scatterplot for LOP as well.

leburnett committed:
remove threshold from scatterplot gen

Now, any cell type with a single column occupied will have a marker in the scatterplot.

floesche committed:
Merge branch 'main' into issue_78_part_2

floesche committed:
minor formatting fixes

leburnett committed:
Merge branch 'main' into issue_78_part_2

leburnett committed:
update tooltip box, hardcode axes

leburnett committed:
increase tooltip font size

floesche committed:
Merge branch 'main' into issue_78_part_2

floesche committed:
use CNS, update dependencies

floesche committed:
add scatter plots per hemisphere

floesche committed:
better tooltips on L / R scatter plots

floesche committed:
update license

floesche committed:
improve layout for scatterplots

floesche committed:
add download button for scatterplots

floesche commented:
I added the scatter plots for L and R and did some layout changes. Let me know if you think they are improvements or if we should go back.
Also, the scale for the coverage factors goes from "n" to ">n" (e.g. 10 and >10 in the screenshot above). What does that mean?

floesche committed:
add subset-medium-optic as a test set

floesche committed:
before, LOVC23 was higlighted when I really just wanted C2

floesche commented:
Re
>  Also, the scale for the coverage factors goes from "n" to ">n" (e.g. 10 and >10 in the screenshot above). What does that mean?

I added a "subset-medium-optic" test set, and there it looks much better. Maybe not worth addressing that point…

floesche committed:
save space in eyemaps

floesche committed:
remove threshold

floesche commented:
I had noticed, that some cells had innervation in some ROI (e.g. C2 in LO), but the data still didn't show up in the plots. It turns out there was another thresholding in the code. Instead of just deleting it, I pulled it out to the config file (see `min_col_count_theshold` from 21434ede3ccde87cacad7505491d16176e1a454a onwards) but set it to 0 for CNS. We can also remove it in the future, if that's not useful.

leburnett commented:
> I added the scatter plots for L and R and did some layout changes. Let me know if you think they are improvements or if we should go back. Also, the scale for the coverage factors goes from "n" to ">n" (e.g. 10 and >10 in the screenshot above). What does that mean?

Yes, I see your comment below this one as well. I think that's just a bug of only having one data point in the plot. I believe the 98th percentile value is used as the colour scale max and then the tick mark is changed to be ">cmax". I think if someone makes a single webpage and wants to look at these plots then they'll have the tooltip box information to help them understand the colour scale.

leburnett commented:
> I had noticed, that some cells had innervation in some ROI (e.g. C2 in LO), but the data still didn't show up in the plots. It turns out there was another thresholding in the code. Instead of just deleting it, I pulled it out to the config file (see `min_col_count_theshold` from [21434ed](https://github.com/reiserlab/neuView/commit/21434ede3ccde87cacad7505491d16176e1a454a) onwards) but set it to 0 for CNS. We can also remove it in the future, if that's not useful.

Thanks for catching this.

floesche commented:
> > I had noticed, that some cells had innervation in some ROI (e.g. C2 in LO), but the data still didn't show up in the plots. It turns out there was another thresholding in the code. Instead of just deleting it, I pulled it out to the config file (see `min_col_count_theshold` from [21434ed](https://github.com/reiserlab/neuView/commit/21434ede3ccde87cacad7505491d16176e1a454a) onwards) but set it to 0 for CNS. We can also remove it in the future, if that's not useful.
> 
> Thanks for catching this.

It would simplify the code and configuration if this wasn't a parameter and we just always show what's in the database… Let me know what you think.

leburnett committed:
Remove min_col_count_threshold

floesche committed:
prevent KeyError/TypeError in _extract_plot_data on partial cache

floesche committed:
handle empty point sets in create-scatter

floesche committed:
Wire scatter service through service container's config

floesche committed:
Wire scatter service through service container's config

floesche committed:
Stop ScatterplotService from mutating cached spatial_metrics

floesche committed:
Stop ScatterplotService from mutating cached spatial_metrics

floesche committed:
Merge branch 'main' into issue_78_part_2.4

floesche committed:
merge main

floesche committed:
merge main

floesche committed:
Merge branch 'main' into issue_78_part_2

floesche committed:
update pixi

floesche committed:
Merge branch 'main' into issue_78_part_2.2

floesche committed:
fix pixi

floesche committed:
Merge branch 'issue_78_part_2' into issue_78_part_2.2

floesche committed:
fix lock

floesche committed:
fix lock

floesche committed:
Merge branch 'issue_78_part_2' into issue_78_part_2.4

floesche committed:
return Result from create_scatterplots, exit non-zero on Err

floesche committed:
return Result from create_scatterplots, exit non-zero on Err

floesche committed:
make ScatterConfig fields actually configurable

floesche committed:
make ScatterConfig fields actually configurable

floesche committed:
hoist axis range to ScatterConfig and skip out-of-range points

floesche committed:
speed up jinja by discovering env only once

floesche committed:
Merge branch 'main' into issue_78_part_2

leburnett committed:
Merge pull request #149 from reiserlab/issue_78_part_2.1

fix create-scatter crash when cached spatial_metrics is missing or None

leburnett committed:
Merge branch 'issue_78_part_2' into issue_78_part_2.2

leburnett committed:
update the scatterplot service test

leburnett committed:
Merge pull request #150 from reiserlab/issue_78_part_2.2

fix create-scatter crash on (side, region) plots with no points

floesche committed:
Merge branch 'issue_78_part_2' into issue_78_part_2.4

floesche committed:
fix incorrect merge

floesche committed:
fix incorrect merge

floesche committed:
Merge branch 'issue_78_part_2' into issue_78_part_2.7

leburnett committed:
Merge pull request #151 from reiserlab/issue_78_part_2.3

Wire scatter service through service container's config

leburnett committed:
Merge pull request #152 from reiserlab/issue_78_part_2.4

Stop ScatterplotService from mutating cached spatial_metrics

leburnett committed:
Merge branch 'issue_78_part_2' into issue_78_part_2.5

leburnett committed:
Merge pull request #154 from reiserlab/issue_78_part_2.5

return Result from create_scatterplots, exit non-zero on Err

leburnett committed:
Merge pull request #155 from reiserlab/issue_78_part_2.6

make ScatterConfig fields actually configurable

leburnett committed:
Add fix from PR150 back in

floesche committed:
combine PRs

floesche committed:
Merge branch 'issue_78_part_2' into issue_78_part_2.7

floesche committed:
compute 'both' side cell count from left+right, drop unreachable else

floesche committed:
wire CreateScatterCommand through CLI to ScatterplotService

floesche committed:
include M neurons

For the unlikely case that any "M" neuron ever reaches this code path,
make sure it doesn't lead to an unexpected result.

floesche committed:
make sure results stay the same for L R both only

floesche committed:
fix typo

floesche committed:
move scatter dir mkdir out of __init__ into create_scatterplots

floesche committed:
extract NeuronTypeDiscovery from IndexService

floesche committed:
scatter polish: drop debug logs, stale docs, and CLI filename duplication; expand tests

floesche committed:
Merge branch 'main' into issue_78_part_2

floesche committed:
cleanup

floesche committed:
Merge branch 'issue_78_part_2' into issue_78_part_2.7

floesche committed:
cleanup previous tests

floesche committed:
fix lock file issue

leburnett committed:
Consolidate scatterplot service tests into a single file

Merge test/test_scatterplot_axis_range.py and
test/services/test_scatterplot_service.py into test/test_scatterplot_service.py.

Both files were testing the same class; the parallel test_scatterplot_service.py
filename in test/ and test/services/ caused merge conflicts during the
#149/#150 stack and was a maintenance hazard. Consolidating removes
that risk and puts all ScatterplotService tests in one discoverable place.

Adds coverage for:
- PR #156: ScatterConfig.axis_min/axis_max + out-of-range filter
- The two-line empty-input guard that protects against ValueError on
  min(coverages) when points is [] or fully filtered out

leburnett committed:
Merge pull request #156 from reiserlab/issue_78_part_2.7

hoist axis range to ScatterConfig and skip out-of-range points

leburnett committed:
Merge branch 'issue_78_part_2' into issue_78_part_2.8

floesche committed:
Merge branch 'issue_78_part_2' into issue_78_part_2.10

leburnett committed:
Merge pull request #157 from reiserlab/issue_78_part_2.8

speed up jinja

leburnett committed:
Merge pull request #158 from reiserlab/issue_78_part_2.10

wire CreateScatterCommand through CLI to ScatterplotService

leburnett committed:
Merge branch 'issue_78_part_2' into issue_78_part_2.9

floesche committed:
adding regression tests

leburnett committed:
Merge pull request #159 from reiserlab/issue_78_part_2.9

compute 'both' side cell count from left+right, drop unreachable else

leburnett committed:
Merge branch 'issue_78_part_2' into issue_78_part_2.12

leburnett committed:
Merge pull request #160 from reiserlab/issue_78_part_2.12

move scatter dir mkdir out of __init__ into create_scatterplots

leburnett committed:
Merge branch 'issue_78_part_2' into issue_78_part_2.13

leburnett committed:
Merge pull request #161 from reiserlab/issue_78_part_2.13

extract NeuronTypeDiscovery from IndexService

leburnett committed:
Merge branch 'issue_78_part_2' into issue_78_part_2.14

leburnett committed:
Add the math-helpers back in. Lost after merge.

leburnett committed:
Merge pull request #162 from reiserlab/issue_78_part_2.14

scatter polish: drop debug logs, stale docs, and CLI filename duplica…

leburnett committed:
Merge branch 'issue_78_part_2' into issue_78_part_2_cleanup

leburnett committed:
Merge pull request #163 from reiserlab/issue_78_part_2_cleanup

cleanup previous tests

leburnett committed:
Make test_neuprint_connector_initialization a true unit test

leburnett committed:
Merge branch 'main' into issue_78_part_2

floesche committed:
higlight the current type a bit better in the scatter plots

Previously, the higlighted type was bigger than the other types and
shrunk after the first mouseover.

In this patch, I use a yellow stroke color for the current type and add
it on top of all other types, so it will always be in the foreground.
The yellow stroke color of the circle stays when the circle is shrunk.

leburnett committed:
Merge pull request #165 from reiserlab/higlight-type

higlight the current type a bit better in the scatter plots
